pkg/tracing/tracing.go

```diff
@@ -0,0 +1,132 @@
+		opts ...grpc.CallOption) (grpc.ClientStream, error) {
+
+		var span ot.Span
+		if p := ot.SpanFromContext(ctx); p != nil {
+			span = tracer.StartSpan(method, ext.SpanKindRPCClient, ot.ChildOf(p.Context()))
+		} else {
+			span = tracer.StartSpan(method, ext.SpanKindRPCClient)
+		}
+		defer span.Finish()
+
+		ctx = ot.ContextWithSpan(ctx, span)
+		ctx = propagateSpan(ctx, tracer, span)
+		return streamer(ctx, desc, cc, method, opts...)
+	}
+}
+
+// CurrentSpan extracts the current span from the context, or returns a no-op span if no span exists in the context.
+// This avoids boilerplate nil checking when calling opentracing.SpanFromContext.
+func CurrentSpan(ctx context.Context) ot.Span {
+	// TODO: figure out how to handle a span not existing in the context: this shouldn't happen because a root trace
+	// is always created for the life of a stream. (Possible cause would be someone creating a new context, or not
+	// propagating the context correctly.)
+	if current := ot.SpanFromContext(ctx); current != nil {
+		return current
+	}
+	return noopSpan
+}
+
+type rootSpanKey struct{}
+
+// RootSpan retrieves the root span from the context. This may be different than the current context returned by
+// opentracing.SpanFromContext(ctx)
+func RootSpan(ctx context.Context) ot.Span {
+	val := ctx.Value(rootSpanKey{})
+	if root, ok := val.(ot.Span); ok {
+		return root
+	}
+	return noopSpan
+}
+
+// StartRootSpan creates a span that is the root of all Istio spans in the current request context. This span will be a
+// child of any spans propagated to the server in the request's metadata. The returned span is retrievable from the
+// context via tracing.RootSpan.
+func StartRootSpan(ctx context.Context, operationName string) (ot.Span, context.Context) {
+	md, ok := metadata.FromContext(ctx)
+	if !ok {
+		md = metadata.New(nil)
+	}
+	spanContext, err := ot.GlobalTracer().Extract(ot.TextMap, metadataReaderWriter{md})
+	if err != nil {
+		glog.Warningf("Failed to extract opentracing metadata with tracer %v from ctx %v with err %v", ot.GlobalTracer(), ctx, err)
+
+		// We set the spancontext to nil if there's an error so we don't get funky values from ext.RPCServerOption
+		// (in particular the mock tracer impl doesn't handle a non nil empty span context gracefully).
+		spanContext = nil
+	}
+	span, ctx := ot.StartSpanFromContext(ctx, operationName, ext.RPCServerOption(spanContext), gRPCComponentTag)
+	ctx = context.WithValue(ctx, rootSpanKey{}, span)
+	return span, ctx
+}
+
+// propagateSpan inserts metadata about the span into the context's metadata so that the span is propagated to the receiver.
+// This should be used to prepare the context for outgoing calls.
+//
+// TODO: consider creating a public version of this method for use at individual call sites if the interceptor isn't
+// sufficient for some reason.
+func propagateSpan(ctx context.Context, tracer ot.Tracer, span ot.Span) context.Context {
+	md, ok := metadata.FromContext(ctx)
+	if !ok {
```

Comment:
Refactored this into a method since it's repeated. Side effect is these `if`s go away.